util/topsql/reporter/subscriber.go

```diff
@@ -0,0 +1,291 @@
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tipb/go-tipb"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// TopSQLPublisher implements TopSQLPublisher.
+//
+// If a client subscribes to TopSQL records, the TopSQLPublisher is responsible for registering them to the reporter.
+// Then the reporter sends data to the client periodically.
+type TopSQLPublisher struct {
+	decodePlan     planBinaryDecodeFunc
+	clientRegistry *ReportClientRegistry
+}
+
+// NewTopSQLPublisher creates a new TopSQLPublisher.
+func NewTopSQLPublisher(
+	decodePlan planBinaryDecodeFunc,
+	clientRegistry *ReportClientRegistry,
+) *TopSQLPublisher {
+	return &TopSQLPublisher{
+		decodePlan:     decodePlan,
+		clientRegistry: clientRegistry,
+	}
+}
+
+var _ tipb.TopSQLPubSubServer = &TopSQLPublisher{}
+
+// Subscribe registers clients to the reporter and redirects data received from reporter
+// to subscribers associated with those clients.
+func (t *TopSQLPublisher) Subscribe(
+	_ *tipb.TopSQLSubRequest,
+	stream tipb.TopSQLPubSub_SubscribeServer,
+) error {
+	sc := newSubClient(stream, t.decodePlan)
+
+	t.clientRegistry.register(sc)
+	sc.run()
+
+	return nil
+}
+
+type subClient struct {
+	stream     tipb.TopSQLPubSub_SubscribeServer
+	sendTaskCh chan sendTask
+	isDown     *atomic.Bool
+
+	decodePlan planBinaryDecodeFunc
+}
+
+func newSubClient(
+	stream tipb.TopSQLPubSub_SubscribeServer,
+	decodePlan planBinaryDecodeFunc,
+) *subClient {
+	return &subClient{
+		stream:     stream,
+		sendTaskCh: make(chan sendTask),
+		isDown:     atomic.NewBool(false),
+
+		decodePlan: decodePlan,
+	}
+}
+
+func (s *subClient) run() {
+	defer s.isDown.Store(true)
+
+	for task := range s.sendTaskCh {
+		ctx, cancel := context.WithTimeout(context.Background(), task.timeout)
+		start := time.Now()
+
+		var err error
+		doneCh := make(chan struct{})
+		go func(task sendTask) {
+			util.WithRecovery(func() {
+				defer func() {
+					doneCh <- struct{}{}
+				}()
+				err = s.doSend(ctx, task.data)
+				if err != nil {
+					reportAllDurationFailedHistogram.Observe(time.Since(start).Seconds())
+				} else {
+					reportAllDurationSuccHistogram.Observe(time.Since(start).Seconds())
+				}
+			}, nil)
+		}(task)
+
+		select {
+		case <-doneCh:
+			cancel()
+			if err != nil {
+				logutil.BgLogger().Warn(
+					"[top-sql] client failed to send data to subscriber",
+					zap.Error(err),
+				)
+				return
+			}
+		case <-ctx.Done():
+			cancel()
+			logutil.BgLogger().Warn(
+				"[top-sql] client failed to send data to subscriber due to timeout",
+				zap.Duration("timeout", task.timeout),
+			)
+			return
+		}
+	}
+}
+
+func (s *subClient) doSend(ctx context.Context, data reportData) error {
+	if err := s.sendCPUTime(ctx, data.collectedData); err != nil {
+		return err
+	}
+	if err := s.sendSQLMeta(ctx, data.normalizedSQLMap); err != nil {
+		return err
+	}
+	return s.sendPlanMeta(ctx, data.normalizedPlanMap)
```

Comment:
in receiver client, they are sent in parallel. In Pub/Sub client, they are sent in serial. Will there be problems? Should we subscribe them independently?

Reply:
Sending in parallel may reduce latency, but will increase resource usage and management difficulty. I don't think it's necessary.